From 28686d9071a67c02c0c67ff4dec07fe137e0a15d Mon Sep 17 00:00:00 2001
From: "mujiburrahimank@maxlinear.com"
 <mujiburrahimank@lvdocker7.lvsw.maxlinear.com>
Date: Sun, 11 Jan 2026 21:55:02 -0800
Subject: [PATCH] WLANRTSYS-100609 M/SCS Northbound APIs & Deny List

support of Northbound APIs for SCS/MSCS to add rules via hostapd_cli or ctrl_iface
Below is the list of supported APIs
SCS APIs
scs_add, scs_rm, scs_list, scs_deny_mac_list, scs_show_deny_mac_list.
MSCS APIs
mscs_add, mscs_rm, mscs_list, mscs_deny_mac_list, show_mscs_deny_mac_list
---
 hostapd/mxl_config.c      | 123 ++++++-
 hostapd/mxl_ctrl_iface.c  | 271 ++++++++++++++-
 hostapd/mxl_hostapd_cli.c | 196 +++++++++++
 hostapd/mxl_hostapd_cli.h |  30 ++
 src/ap/mxl_config.h       |  18 +
 src/ap/mxl_hostapd.h      |  25 ++
 src/ap/mxl_ieee802_11.c   |  61 ++++
 src/ap/mxl_ieee802_11.h   |   3 +
 src/ap/mxl_wmm.c          | 680 +++++++++++++++++++++++++++++++++++++-
 src/ap/mxl_wmm.h          |  14 +-
 src/ap/sta_info.c         |  15 +-
 11 files changed, 1414 insertions(+), 22 deletions(-)

diff --git a/hostapd/mxl_config.c b/hostapd/mxl_config.c
index d11b0c19f..906d08978 100644
--- a/hostapd/mxl_config.c
+++ b/hostapd/mxl_config.c
@@ -882,6 +882,120 @@ static int mxl_hostapd_parse_colocated_6g_radio_info(struct hostapd_config *conf
 	return 0;
 }
 
+#if defined(CONFIG_SCS_MXL) || defined(CONFIG_MSCS_MXL)
+int mxl_parse_mac_str_to_hex(const char *s, u8 mac[ETH_ALEN])
+{
+	unsigned int o[ETH_ALEN];
+	if (sscanf(s, "%02x:%02x:%02x:%02x:%02x:%02x",
+				&o[0], &o[1], &o[2], &o[3], &o[4], &o[5]) != 6)
+		return -1;
+	for (int i = 0; i < ETH_ALEN; i++)
+		mac[i] = (u8) o[i];
+	return 0;
+}
+
+static int mxl_hostapd_parse_qos_deny_mac_list(struct mxl_qos_deny_list **qos_head, char *mac_list)
+{
+	u8 sta_mac[ETH_ALEN];
+	struct mxl_qos_deny_list *qos_free = *qos_head, *tmp = NULL;
+	char *saveptr = NULL, *mac = NULL;
+	char *tok = strtok_r(mac_list, " ", &saveptr);
+
+	while (qos_free) {
+		tmp = qos_free->next;
+		os_free(qos_free);
+		qos_free = tmp;
+	}
+	*qos_head = NULL;
+
+	if (!tok) {
+		wpa_printf(MSG_DEBUG, "(M)SCS DENY MAC list empty");
+		return 0;
+	}
+
+	while (tok) {
+		mac = tok;
+		if (mac[0] != '\0') {
+			if (mxl_parse_mac_str_to_hex(mac, sta_mac) != 0) {
+				wpa_printf(MSG_DEBUG, "(M)SCS DENY MAC Invalid MAC entry");
+				return 0;
+			}
+
+			struct mxl_qos_deny_list *new = (struct mxl_qos_deny_list *)os_malloc(sizeof(struct mxl_qos_deny_list));
+			if (!new) {
+				return 0;
+			}
+			new->next = NULL;
+			os_memcpy(new->mxl_qos_deny_mac, sta_mac, ETH_ALEN);
+
+			if(!tmp)
+				tmp = new;
+			else {
+				new->next = tmp;
+				tmp = new;
+			}
+		}
+		tok = strtok_r(NULL, " ", &saveptr);
+	}
+
+	*qos_head = tmp;
+
+	return 0;
+}
+
+int mxl_config_check_qos_deny_list(struct hostapd_config *conf, const char *buf, char* pos)
+{
+	struct mxl_hostapd_config *mxl_conf = &conf->mxl_conf;
+	if (os_strcmp(buf, "scs_deny_mac") == 0) {
+#ifdef CONFIG_SCS_MXL
+		mxl_hostapd_parse_qos_deny_mac_list(&mxl_conf->scs_deny, pos);
+#endif /* CONFIG_SCS_MXL */
+	}
+	else if (os_strcmp(buf, "mscs_deny_mac") == 0) {
+#ifdef CONFIG_MSCS_MXL
+		mxl_hostapd_parse_qos_deny_mac_list(&mxl_conf->mscs_deny, pos);
+#endif /* CONFIG_MSCS_MXL */
+	}
+	else {
+		return 0;
+	}
+
+	return 1;
+}
+
+void mxl_hostapd_config_free_qos_deny_list(struct hostapd_config *conf)
+{
+	struct mxl_hostapd_config *mxl_conf = &conf->mxl_conf;
+	struct mxl_qos_deny_list *tmp = NULL;
+
+#ifdef CONFIG_SCS_MXL
+	if (mxl_conf->scs_deny) {
+		tmp = mxl_conf->scs_deny;
+		while (tmp) {
+			mxl_conf->scs_deny = tmp->next;
+			os_free(tmp);
+			tmp = mxl_conf->scs_deny;
+		}
+		mxl_conf->scs_deny = NULL;
+	}
+#endif /* CONFIG_SCS_MXL */
+
+#ifdef CONFIG_MSCS_MXL
+	if (mxl_conf->mscs_deny) {
+		tmp = mxl_conf->mscs_deny;
+		while (tmp) {
+			mxl_conf->mscs_deny = tmp->next;
+			os_free(tmp);
+			tmp = mxl_conf->mscs_deny;
+		}
+		mxl_conf->mscs_deny = NULL;
+	}
+#endif /* CONFIG_MSCS_MXL */
+	return;
+}
+
+#endif /* CONFIG_SCS_MXL  || CONFIG_MSCS_MXL */
+
 #ifdef CONFIG_IEEE80211BE
 
 static int
@@ -1722,6 +1836,10 @@ int mxl_hostapd_config_fill (struct hostapd_config *conf,
 	} else if(mxl_config_check_dfs_debug_chan(conf, buf, pos)) {
 		return 1;
 #endif /* CONFIG_VENDOR_MXL_PROPRIETARY */
+#if defined(CONFIG_SCS_MXL) || defined(CONFIG_MSCS_MXL)
+	} else if(mxl_config_check_qos_deny_list(conf, buf, pos)) {
+		return 1;
+#endif /* CONFIG_SCS_MXL  || CONFIG_MSCS_MXL */
 	} else if (os_strcmp(buf, "sub_band_dfs") == 0) {
 		int val = atoi(pos);
 		if(!IS_VALID_RANGE(val, SUBBAND_DFS_MIN, SUBBAND_DFS_MAX)) {
@@ -2177,8 +2295,11 @@ void mxl_hostapd_config_free(struct hostapd_config *conf)
 	struct mxl_hostapd_config *mxl_conf = &conf->mxl_conf;
 
 #ifdef CONFIG_VENDOR_MXL_PROPRIETARY
-        mxl_hostapd_config_free_acs(conf);
+	mxl_hostapd_config_free_acs(conf);
 #endif /* CONFIG_VENDOR_MXL_PROPRIETARY */
+#if defined(CONFIG_SCS_MXL) || defined(CONFIG_MSCS_MXL)
+	mxl_hostapd_config_free_qos_deny_list(conf);
+#endif /* CONFIG_SCS_MXL  || CONFIG_MSCS_MXL */
 
 	os_free(mxl_conf->dfs_channels_state_file);
 }
diff --git a/hostapd/mxl_ctrl_iface.c b/hostapd/mxl_ctrl_iface.c
index 591acd94a..bd92b35e3 100644
--- a/hostapd/mxl_ctrl_iface.c
+++ b/hostapd/mxl_ctrl_iface.c
@@ -3193,7 +3193,129 @@ static int mxl_hostapd_ctrl_iface_scs_unsolicited_rem(struct hostapd_data *hapd,
 
 	return res;
 }
-#endif
+
+static int mxl_hostapd_ctrl_iface_show_scs_deny_mac_list (struct hostapd_data *hapd, char *reply, size_t replylen)
+{
+	int rc = 0;
+	char scs_deny_mac_buf[2048] = {'\0'};
+
+	if (!hapd->conf->mxl_bss_conf.scs_enable) {
+		wpa_printf(MSG_ERROR, "SCS: ctrl-iface show_scs_deny_mac_list - not enabled/supported");
+		return -1;
+	}
+
+	/* get scs deny mac list */
+	rc  = mxl_get_qos_deny_mac_list(hapd, scs_deny_mac_buf, true);
+	if (rc)
+		rc = sprintf_s(reply, replylen, "SCS deny mac list:\n%s", scs_deny_mac_buf);
+	else
+		rc = sprintf_s(reply, replylen, "SCS deny mac list empty\n");
+	return rc;
+}
+
+static int mxl_hostapd_ctrl_iface_scs_deny_mac_list (struct hostapd_data *hapd, char *buf, char *reply, size_t replylen)
+{
+	int rc = 0;
+	char mac_str[2048];
+
+	if (!hapd->conf->mxl_bss_conf.scs_enable) {
+		wpa_printf(MSG_ERROR, "SCS: ctrl-iface scs_deny_mac_list - not enabled/supported");
+		return -1;
+	}
+
+	os_strlcpy(mac_str, buf, sizeof(mac_str));
+	/* set scs deny mac list */
+	rc = mxl_set_qos_deny_mac_list(hapd, mac_str, true);
+	if (rc < 0)
+		return -1;
+	else if (rc == 1)
+		rc = sprintf_s(reply, replylen, "SCS deny mac list were updated\n");
+	else
+		rc = sprintf_s(reply, replylen, "SCS deny mac list is empty\n");
+
+	return rc;
+}
+static int mxl_hostapd_ctrl_iface_scs_list_rule(struct hostapd_data *hapd, char *reply, size_t replylen)
+{
+	int rc = 0;
+	char scs_list_buf[2048] = {'\0'};
+
+	if (!hapd->conf->mxl_bss_conf.scs_enable) {
+		wpa_printf(MSG_ERROR, "SCS: ctrl-iface scs_list - not enabled/supported");
+		return -1;
+	}
+
+	/* Get stored rules in a buf */
+	rc = mxl_northbound_scs_list_rule(hapd, scs_list_buf);
+	if (rc)
+		rc = sprintf_s(reply, replylen, "SCS rules :\n%s", scs_list_buf);
+	else
+		rc = sprintf_s(reply, replylen, "SCS rules empty\n");
+
+	return rc;
+}
+
+static int mxl_hostapd_ctrl_iface_scs_remove_rule(struct hostapd_data *hapd, char *buf, char *reply, size_t replylen)
+{
+	int rc = 0, n_args = 0;
+	unsigned int scsId;
+	char mac_str[64] = {0};
+
+	if (!hapd->conf->mxl_bss_conf.scs_enable) {
+		wpa_printf(MSG_ERROR, "SCS: ctrl-iface scs_remove - not enabled/supported");
+		return -1;
+	}
+
+	n_args = sscanf(buf, " %63s %u", mac_str, &scsId);
+	if (n_args != 2) {
+		wpa_printf(MSG_ERROR, "Parse error: expected 2 tokens, got %d\n", n_args);
+		return -1;
+	}
+
+	/* remove and handle scs northbound request */
+	rc = mxl_northbound_scs_remove_req(hapd, mac_str, scsId);
+
+	if (rc) {
+		rc = sprintf_s(reply, replylen, "SCS rule removed\nSTA_MAC %s  SCS_ID %u\n", mac_str, scsId);
+	}
+	else {
+		rc = sprintf_s(reply, replylen, "Failed to remove SCS rule\nSTA_MAC %s  SCS_ID %u\n", mac_str, scsId);
+	}
+
+	return rc;
+}
+
+static int mxl_hostapd_ctrl_iface_scs_add_rule(struct hostapd_data *hapd, char *buf, char *reply, size_t replylen)
+{
+	int rc = 0, n_args = 0;
+	unsigned int scsId;
+	char mac_str[64] = {0};
+	char hex_str[1024] = {0};
+
+	if (!hapd->conf->mxl_bss_conf.scs_enable) {
+		wpa_printf(MSG_ERROR, "SCS: ctrl-iface scs_add/remove - not enabled/supported");
+		return -1;
+	}
+
+	n_args = sscanf(buf, " %63s %u %[^\n]", mac_str, &scsId, hex_str);
+	if (n_args != 3) {
+		wpa_printf(MSG_ERROR, "Parse error: expected 3 tokens, got %d\n", n_args);
+		return -1;
+	}
+
+	/* store and handle scs northbound request */
+	rc = mxl_northbound_scs_add_req(hapd, mac_str, scsId, hex_str);
+
+	if (rc) {
+		rc = sprintf_s(reply, replylen, "SCS rule added\nSTA_MAC : %s  SCS_ID : %u\n", mac_str, scsId);
+	}
+	else {
+		rc = sprintf_s(reply, replylen, "Failed to add SCS rule\nSTA_MAC %s  SCS_ID : %u\n", mac_str, scsId);
+	}
+
+	return rc;
+}
+#endif /* CONFIG_SCS_MXL */
 
 #ifdef CONFIG_MSCS_MXL
 static int mxl_hostapd_drv_get_mscs_enable(struct hostapd_data *hapd, int *enabled)
@@ -3245,6 +3367,127 @@ static int mxl_hostapd_ctrl_iface_mscs_unsolicited_rem(struct hostapd_data *hapd
 
 	return res;
 }
+
+static int mxl_hostapd_ctrl_iface_show_mscs_deny_mac_list (struct hostapd_data *hapd, char *reply, size_t replylen)
+{
+	int rc = 0;
+	char mscs_deny_mac_buf[2048] = {'\0'};
+
+	if (!hapd->conf->mxl_bss_conf.mscs_enable) {
+		wpa_printf(MSG_ERROR, "MSCS: ctrl-iface show_mscs_deny_mac_list - not enabled/supported");
+		return -1;
+	}
+
+	/* get mscs deny mac list */
+	rc  = mxl_get_qos_deny_mac_list(hapd, mscs_deny_mac_buf, false);
+	if (rc)
+		rc = sprintf_s(reply, replylen, "MSCS deny mac list:\n%s", mscs_deny_mac_buf);
+	else
+		rc = sprintf_s(reply, replylen, "MSCS deny mac list empty\n");
+	return rc;
+}
+
+static int mxl_hostapd_ctrl_iface_mscs_deny_mac_list (struct hostapd_data *hapd, char *buf, char *reply, size_t replylen)
+{
+	int rc = 0;
+	char mac_str[2048];
+
+	if (!hapd->conf->mxl_bss_conf.mscs_enable) {
+		wpa_printf(MSG_ERROR, "MSCS: ctrl-iface mscs_deny_mac_list - not enabled/supported");
+		return -1;
+	}
+
+	os_strlcpy(mac_str, buf, sizeof(mac_str));
+	/* set mscs deny mac list */
+	rc = mxl_set_qos_deny_mac_list(hapd, mac_str, false);
+	if (rc < 0)
+		return -1;
+	else if (rc == 1)
+		rc = sprintf_s(reply, replylen, "MSCS deny mac list were updated\n");
+	else
+		rc = sprintf_s(reply, replylen, "MSCS deny mac list is empty\n");
+
+	return rc;
+}
+
+static int mxl_hostapd_ctrl_iface_mscs_list_rule(struct hostapd_data *hapd, char *reply, size_t replylen)
+{
+	int rc = 0;
+	char mscs_list_buf[2048] = {'\0'};
+
+	if (!hapd->conf->mxl_bss_conf.mscs_enable) {
+		wpa_printf(MSG_ERROR, "MSCS: ctrl-iface mscs_list - not enabled/supported");
+		return -1;
+	}
+
+	/* Get stored rules in a buf */
+	rc = mxl_northbound_mscs_list_rule(hapd, mscs_list_buf);
+	if (rc)
+		rc = sprintf_s(reply, replylen, "MSCS rules :\n%s", mscs_list_buf);
+	else
+		rc = sprintf_s(reply, replylen, "MSCS rules empty\n");
+
+	return rc;
+}
+
+static int mxl_hostapd_ctrl_iface_mscs_remove_rule(struct hostapd_data *hapd, char *buf, char *reply, size_t replylen)
+{
+	int rc = 0, n_args = 0;
+	char mac_str[64] = {0};
+
+	if (!hapd->conf->mxl_bss_conf.mscs_enable) {
+		wpa_printf(MSG_ERROR, "MSCS: ctrl-iface mscs_add/remove - not enabled/supported");
+		return -1;
+	}
+
+	n_args = sscanf(buf, " %63s", mac_str);
+	if (n_args != 1) {
+		wpa_printf(MSG_ERROR, "Parse error: expected 1 tokens, got %d\n", n_args);
+		return -1;
+	}
+
+	/* remove and handle mscs northbound request */
+	rc = mxl_northbound_mscs_remove_req(hapd, mac_str);
+
+	if (rc) {
+		rc = sprintf_s(reply, replylen, "MSCS rule removed\nSTA_MAC %s\n", mac_str);
+	}
+	else {
+		rc = sprintf_s(reply, replylen, "Failed to remove MSCS rule\nSTA_MAC %s\n", mac_str);
+	}
+
+	return rc;
+}
+
+static int mxl_hostapd_ctrl_iface_mscs_add_rule(struct hostapd_data *hapd, char *buf, char *reply, size_t replylen)
+{
+	int rc = 0, n_args = 0;
+	char mac_str[64] = {0};
+	char hex_str[1024] = {0};
+
+	if (!hapd->conf->mxl_bss_conf.mscs_enable) {
+		wpa_printf(MSG_ERROR, "MSCS: ctrl-iface mscs_add/remove - not enabled/supported");
+		return -1;
+	}
+	
+	n_args = sscanf(buf, " %63s %[^\n]", mac_str, hex_str);
+	if (n_args != 2) {
+		wpa_printf(MSG_ERROR, "Parse error: expected 2 tokens, got %d\n", n_args);
+		return -1;
+	}
+
+	/* store and handle mscs northbound request */
+	rc = mxl_northbound_mscs_add_req(hapd, mac_str, hex_str);
+
+	if (rc) {
+		rc = sprintf_s(reply, replylen, "MSCS rule added\nSTA_MAC : %s\n", mac_str);
+	}
+	else {
+		rc = sprintf_s(reply, replylen, "Failed to add MSCS rule\nSTA_MAC %s\n", mac_str);
+	}
+
+	return rc;
+}
 #endif /* CONFIG_MSCS_MXL */
 
 int mxl_hostapd_ctrl_iface_receive_process(struct hostapd_data *hapd,
@@ -3486,6 +3729,19 @@ int mxl_hostapd_ctrl_iface_receive_process(struct hostapd_data *hapd,
 	} else if (os_strncmp(buf, "SCS_REM_REQ ", sizeof("SCS_REM_REQ ") - 1) == 0) {
 		if (mxl_hostapd_ctrl_iface_scs_unsolicited_rem(hapd, buf + sizeof("SCS_REM_REQ ") -1, reply, reply_size) < 0)
 			*reply_len = -1;
+	} else if (os_strncmp(buf, "SCS_ADD_RULE ", 13) == 0) {
+		*reply_len = mxl_hostapd_ctrl_iface_scs_add_rule(hapd, buf + sizeof("SCS_ADD_RULE ") -1,
+			reply, reply_size);
+	} else if (os_strncmp(buf, "SCS_REMOVE_RULE ", 16) == 0) {
+		*reply_len = mxl_hostapd_ctrl_iface_scs_remove_rule(hapd, buf + sizeof("SCS_REMOVE_RULE ") -1,
+			reply, reply_size);
+	} else if (os_strcmp(buf, "SCS_LIST_RULE") == 0) {
+		*reply_len = mxl_hostapd_ctrl_iface_scs_list_rule(hapd, reply, reply_size);
+	} else if (os_strncmp(buf, "SCS_DENY_MAC ", 13) == 0) {
+		*reply_len = mxl_hostapd_ctrl_iface_scs_deny_mac_list(hapd, buf + sizeof("SCS_DENY_MAC ") -1,
+			reply, reply_size);
+	} else if (os_strcmp(buf, "SHOW_SCS_DENY_MAC") == 0) {
+		*reply_len = mxl_hostapd_ctrl_iface_show_scs_deny_mac_list(hapd, reply, reply_size);
 #endif /* CONFIG_SCS */
 #ifdef CONFIG_MSCS_MXL
 	} else if (os_strncmp(buf, "GET_MSCS_ENABLE", 15) == 0) {
@@ -3494,6 +3750,19 @@ int mxl_hostapd_ctrl_iface_receive_process(struct hostapd_data *hapd,
 	} else if (os_strncmp(buf, "MSCS_REM_REQ ", sizeof("MSCS_REM_REQ ") - 1) == 0) {
 		if (mxl_hostapd_ctrl_iface_mscs_unsolicited_rem(hapd, buf + sizeof("MSCS_REM_REQ ") -1, reply, reply_size) < 0)
 			*reply_len = -1;
+	} else if (os_strncmp(buf, "MSCS_ADD_RULE ", 14) == 0) {
+		*reply_len = mxl_hostapd_ctrl_iface_mscs_add_rule(hapd, buf + sizeof("MSCS_ADD_RULE ") -1,
+			reply, reply_size);
+	} else if (os_strncmp(buf, "MSCS_REMOVE_RULE ", 17) == 0) {
+		*reply_len = mxl_hostapd_ctrl_iface_mscs_remove_rule(hapd, buf + sizeof("MSCS_REMOVE_RULE ") -1,
+			reply, reply_size);
+	} else if (os_strcmp(buf, "MSCS_LIST_RULE") == 0) {
+		*reply_len = mxl_hostapd_ctrl_iface_mscs_list_rule(hapd, reply, reply_size);
+	} else if (os_strncmp(buf, "MSCS_DENY_MAC ", 13) == 0) {
+		*reply_len = mxl_hostapd_ctrl_iface_mscs_deny_mac_list(hapd, buf + sizeof("MSCS_DENY_MAC ") -1,
+			reply, reply_size);
+	} else if (os_strcmp(buf, "SHOW_MSCS_DENY_MAC") == 0) {
+		*reply_len = mxl_hostapd_ctrl_iface_show_mscs_deny_mac_list(hapd, reply, reply_size);
 #endif /* CONFIG_MSCS */
 	} else {
 		return 0;
diff --git a/hostapd/mxl_hostapd_cli.c b/hostapd/mxl_hostapd_cli.c
index 85afe5cbb..06300952c 100644
--- a/hostapd/mxl_hostapd_cli.c
+++ b/hostapd/mxl_hostapd_cli.c
@@ -605,3 +605,199 @@ int mxl_hostapd_cli_cmd_mscs_rem(struct wpa_ctrl *ctrl, int argc, char *argv[])
 	}
 	return mxl_wpa_ctrl_command(ctrl, cmd);
 }
+
+int mxl_hostapd_cli_cmd_scs_add(struct wpa_ctrl *ctrl, int argc,
+					char *argv[])
+{
+	char cmd[2048];
+	int res = 0;
+
+	if (argc < 3) {
+		printf("Usage: scs_add <STA_MAC> <SCS_ID> <hexdump>\n");
+		return -1;
+	}
+
+	res = sprintf_s(cmd, sizeof(cmd), "SCS_ADD_RULE %s %s %s",
+			argv[0], argv[1], argv[2]);
+
+	if (res <= 0) {
+		printf("%s: sprintf_s error %d", __func__, res);
+		return -1;
+	}
+
+	return mxl_wpa_ctrl_command(ctrl, cmd);
+}
+
+int mxl_hostapd_cli_cmd_scs_remove(struct wpa_ctrl *ctrl, int argc,
+					char *argv[])
+{
+	char cmd[2048];
+	int res = 0;
+
+	if (argc < 2) {
+		printf("Usage: scs_rm <STA_MAC> <SCS_ID>\n");
+		return -1;
+	}
+
+	res = sprintf_s(cmd, sizeof(cmd), "SCS_REMOVE_RULE %s %s",
+			argv[0], argv[1]);
+
+	if (res <= 0) {
+		printf("%s: sprintf_s error %d", __func__, res);
+		return -1;
+	}
+
+	return mxl_wpa_ctrl_command(ctrl, cmd);
+}
+
+int mxl_hostapd_cli_cmd_scs_list(struct wpa_ctrl *ctrl, int argc,
+					char *argv[])
+{
+	char cmd[64];
+	int res = 0;
+
+	res = sprintf_s(cmd, sizeof(cmd), "SCS_LIST_RULE");
+
+	if (res <= 0) {
+		printf("%s: sprintf_s error %d", __func__, res);
+		return -1;
+	}
+
+	return mxl_wpa_ctrl_command(ctrl, cmd);
+}
+
+int mxl_hostapd_cli_cmd_mscs_add(struct wpa_ctrl *ctrl, int argc,
+					char *argv[])
+{
+	char cmd[2048];
+	int res = 0;
+
+	if (argc < 2) {
+		printf("Usage: mscs_add <STA_MAC> <hexdump>\n");
+		return -1;
+	}
+
+	res = sprintf_s(cmd, sizeof(cmd), "MSCS_ADD_RULE %s %s",
+			argv[0], argv[1]);
+
+	if (res <= 0) {
+		printf("%s: sprintf_s error %d", __func__, res);
+		return -1;
+	}
+
+	return mxl_wpa_ctrl_command(ctrl, cmd);
+}
+
+int mxl_hostapd_cli_cmd_mscs_remove(struct wpa_ctrl *ctrl, int argc,
+					char *argv[])
+{
+	char cmd[2048];
+	int res = 0;
+
+	if (argc < 1) {
+		printf("Usage: mscs_rm <STA_MAC>\n");
+		return -1;
+	}
+
+	res = sprintf_s(cmd, sizeof(cmd), "MSCS_REMOVE_RULE %s",
+			argv[0]);
+
+	if (res <= 0) {
+		printf("%s: sprintf_s error %d", __func__, res);
+		return -1;
+	}
+
+	return mxl_wpa_ctrl_command(ctrl, cmd);
+}
+
+int mxl_hostapd_cli_cmd_mscs_list(struct wpa_ctrl *ctrl, int argc,
+					char *argv[])
+{
+	char cmd[64];
+	int res = 0;
+
+	res = sprintf_s(cmd, sizeof(cmd), "MSCS_LIST_RULE");
+
+	if (res <= 0) {
+		printf("%s: sprintf_s error %d", __func__, res);
+		return -1;
+	}
+
+	return mxl_wpa_ctrl_command(ctrl, cmd);
+}
+
+int mxl_hostapd_cli_cmd_scs_deny_mac_list(struct wpa_ctrl *ctrl, int argc,
+					char *argv[])
+{
+	char cmd[2048];
+	int res = 0;
+
+	if (argc < 1) {
+		printf("Usage: scs_deny_mac_list <MAC1 MAC2 MAC3...>\n");
+		return -1;
+	}
+
+	res = sprintf_s(cmd, sizeof(cmd), "SCS_DENY_MAC %s",
+			argv[0]);
+
+	if (res <= 0) {
+		printf("%s: sprintf_s error %d", __func__, res);
+		return -1;
+	}
+
+	return mxl_wpa_ctrl_command(ctrl, cmd);
+}
+
+int mxl_hostapd_cli_cmd_mscs_deny_mac_list(struct wpa_ctrl *ctrl, int argc,
+					char *argv[])
+{
+	char cmd[2048];
+	int res = 0;
+
+	if (argc < 1) {
+		printf("Usage: mscs_deny_mac_list <MAC1 MAC2 MAC3...>\n");
+		return -1;
+	}
+
+	res = sprintf_s(cmd, sizeof(cmd), "MSCS_DENY_MAC %s",
+			argv[0]);
+
+	if (res <= 0) {
+		printf("%s: sprintf_s error %d", __func__, res);
+		return -1;
+	}
+
+	return mxl_wpa_ctrl_command(ctrl, cmd);
+}
+
+int mxl_hostapd_cli_cmd_show_scs_deny_mac_list(struct wpa_ctrl *ctrl, int argc,
+					char *argv[])
+{
+	char cmd[64];
+	int res = 0;
+
+	res = sprintf_s(cmd, sizeof(cmd), "SHOW_SCS_DENY_MAC");
+
+	if (res <= 0) {
+		printf("%s: sprintf_s error %d", __func__, res);
+		return -1;
+	}
+
+	return mxl_wpa_ctrl_command(ctrl, cmd);
+}
+
+int mxl_hostapd_cli_cmd_show_mscs_deny_mac_list(struct wpa_ctrl *ctrl, int argc,
+					char *argv[])
+{
+	char cmd[64];
+	int res = 0;
+
+	res = sprintf_s(cmd, sizeof(cmd), "SHOW_MSCS_DENY_MAC");
+
+	if (res <= 0) {
+		printf("%s: sprintf_s error %d", __func__, res);
+		return -1;
+	}
+
+	return mxl_wpa_ctrl_command(ctrl, cmd);
+}
diff --git a/hostapd/mxl_hostapd_cli.h b/hostapd/mxl_hostapd_cli.h
index f1cb73b1d..016b75083 100644
--- a/hostapd/mxl_hostapd_cli.h
+++ b/hostapd/mxl_hostapd_cli.h
@@ -112,10 +112,30 @@ int mxl_wpa_ctrl_command(struct wpa_ctrl *ctrl, const char *cmd);
 	  " = Get scs enable/disable status"}, \
 	{"scs_remove", mxl_hostapd_cli_cmd_scs_rem, NULL, \
 	  " = Unsolicited scs rem req"}, \
+	{ "scs_add", mxl_hostapd_cli_cmd_scs_add, NULL, \
+     "<mac_addr> <scs_id> <qos_des_hexdump> = scs rule will be added hapd to driver" }, \
+	{ "scs_rm", mxl_hostapd_cli_cmd_scs_remove, NULL, \
+     "<mac_addr> <scs_id> = scs rule will be removed from hapd as well as driver" }, \
+	{ "scs_list", mxl_hostapd_cli_cmd_scs_list, NULL, \
+     "= scs rule will be listed" }, \
+	{ "show_scs_deny_mac_list", mxl_hostapd_cli_cmd_show_scs_deny_mac_list, NULL, \
+     "= show list of station mac which are denied for scs" }, \
+	{ "scs_deny_mac_list", mxl_hostapd_cli_cmd_scs_deny_mac_list, NULL, \
+     "<mac_addr1 mac_addr2..> = list of station mac which are denied for scs" }, \
 	{ "get_mscs_enable", mxl_hostapd_cli_cmd_get_mscs_enable, NULL, \
 	  " = Get mscs enable/disable status"}, \
 	{ "mscs_remove", mxl_hostapd_cli_cmd_mscs_rem, NULL, \
 	  " = Unsolicited mscs rem req"}, \
+	{ "mscs_add", mxl_hostapd_cli_cmd_mscs_add, NULL, \
+     "<mac_addr> <qos_des_hexdump> = mscs rule will be added hapd to driver" }, \
+	{ "mscs_rm", mxl_hostapd_cli_cmd_mscs_remove, NULL, \
+     "<mac_addr> = mscs rule will be removed from hapd as well as driver" }, \
+	{ "mscs_list", mxl_hostapd_cli_cmd_mscs_list, NULL, \
+     "= mscs rule will be listed" }, \
+	{ "mscs_deny_mac_list", mxl_hostapd_cli_cmd_mscs_deny_mac_list, NULL, \
+     "<mac_addr1 mac_addr2..> = list of station mac which are denied for mscs" }, \
+	{ "show_mscs_deny_mac_list", mxl_hostapd_cli_cmd_show_mscs_deny_mac_list, NULL, \
+     "= show list of station mac which are denied for mscs" }, \
 
 #ifdef CONFIG_ACS
 #define MXL_CLI_COMMANDS_ACS \
@@ -241,6 +261,16 @@ int mxl_hostapd_cli_cmd_get_scs_enable(struct wpa_ctrl *ctrl, int argc, char *ar
 int mxl_hostapd_cli_cmd_scs_rem(struct wpa_ctrl *ctrl, int argc, char *argv[]);
 int mxl_hostapd_cli_cmd_get_mscs_enable(struct wpa_ctrl *ctrl, int argc, char *argv[]);
 int mxl_hostapd_cli_cmd_mscs_rem(struct wpa_ctrl *ctrl, int argc, char *argv[]);
+int mxl_hostapd_cli_cmd_scs_add(struct wpa_ctrl *ctrl, int argc, char *argv[]);
+int mxl_hostapd_cli_cmd_scs_remove(struct wpa_ctrl *ctrl, int argc, char *argv[]);
+int mxl_hostapd_cli_cmd_scs_list(struct wpa_ctrl *ctrl, int argc, char *argv[]);
+int mxl_hostapd_cli_cmd_mscs_add(struct wpa_ctrl *ctrl, int argc, char *argv[]);
+int mxl_hostapd_cli_cmd_mscs_remove(struct wpa_ctrl *ctrl, int argc, char *argv[]);
+int mxl_hostapd_cli_cmd_mscs_list(struct wpa_ctrl *ctrl, int argc, char *argv[]);
+int mxl_hostapd_cli_cmd_scs_deny_mac_list(struct wpa_ctrl *ctrl, int argc, char *argv[]);
+int mxl_hostapd_cli_cmd_mscs_deny_mac_list(struct wpa_ctrl *ctrl, int argc, char *argv[]);
+int mxl_hostapd_cli_cmd_show_scs_deny_mac_list(struct wpa_ctrl *ctrl, int argc, char *argv[]);
+int mxl_hostapd_cli_cmd_show_mscs_deny_mac_list(struct wpa_ctrl *ctrl, int argc, char *argv[]);
 
 
 
diff --git a/src/ap/mxl_config.h b/src/ap/mxl_config.h
index 695e6fbd0..edef7fa27 100644
--- a/src/ap/mxl_config.h
+++ b/src/ap/mxl_config.h
@@ -355,6 +355,13 @@ typedef u8 macaddr[ETH_ALEN];
 #define PROBE_REQUEST_TIME_MIN 0
 #define PROBE_REQUEST_TIME_MAX 60
 
+#if defined(CONFIG_SCS_MXL) || defined(CONFIG_MSCS_MXL)
+struct mxl_qos_deny_list {
+	u8 mxl_qos_deny_mac[ETH_ALEN];
+	struct mxl_qos_deny_list *next;
+};
+#endif /* CONFIG_SCS_MXL  || CONFIG_MSCS_MXL */
+
 /**
  * struct mxl_hostapd_config - Per-radio interface configuration
  */
@@ -469,6 +476,12 @@ struct mxl_hostapd_config {
 #ifdef CONFIG_VENDOR_MXL_PROPRIETARY
 	int dynamic_wmm;
 #endif /* CONFIG_VENDOR_MXL_PROPRIETARY */
+#ifdef CONFIG_SCS_MXL
+	struct mxl_qos_deny_list *scs_deny;
+#endif /* CONFIG_SCS_MXL */
+#ifdef CONFIG_MSCS_MXL
+	struct mxl_qos_deny_list *mscs_deny;
+#endif /* CONFIG_MSCS_MXL */
 };
 
 enum duplicate_beacon_bw {
@@ -559,6 +572,11 @@ int mxl_parse_wpabuf_hex(int line, const char *name, struct wpabuf **buf, const
 int mxl_config_check_dynamic_wmm_wrap(struct hostapd_config *conf, const char *buf, char* pos);
 int mxl_hostapd_ctrl_iface_set_wmm_wrap(struct hostapd_data *hapd, char *cmd, int *ret);
 bool mxl_hostapd_event_ltq_dynamic_wmm(enum wpa_event_type event, struct hostapd_data *hapd, union wpa_event_data *data);
+#if defined(CONFIG_SCS_MXL) || defined(CONFIG_MSCS_MXL)
+int mxl_config_check_qos_deny_list(struct hostapd_config *conf, const char *buf, char* pos);
+void mxl_hostapd_config_free_qos_deny_list(struct hostapd_config *conf);
+int mxl_parse_mac_str_to_hex(const char *s, u8 mac[ETH_ALEN]);
+#endif /* CONFIG_SCS_MXL  || CONFIG_MSCS_MXL */
 
 
 
diff --git a/src/ap/mxl_hostapd.h b/src/ap/mxl_hostapd.h
index e222ba0c6..dc8e7c7a6 100644
--- a/src/ap/mxl_hostapd.h
+++ b/src/ap/mxl_hostapd.h
@@ -198,6 +198,25 @@ struct event_info {
 	char msg_string[MAX_EVENT_MSG_LEN];
 };
 
+#ifdef CONFIG_SCS_MXL
+struct mxl_scs_entry {
+	unsigned int scsId;
+	u8 sta_mac[ETH_ALEN];
+	u8 *mgmt_bytes;  /* radiotap-stripped 802.11 bytes */
+	size_t mgmt_len;
+	struct mxl_scs_entry *next;
+};
+#endif /* CONFIG_SCS_MXL */
+
+#ifdef CONFIG_MSCS_MXL
+struct mxl_mscs_entry {
+	u8 sta_mac[ETH_ALEN];
+	u8 *mgmt_bytes;  /* radiotap-stripped 802.11 bytes */
+	size_t mgmt_len;
+	struct mxl_mscs_entry *next;
+};
+#endif /* CONFIG_MSCS_MXL */
+
 struct mxl_hostapd_data {
 	enum hostapd_channel_switch_reason {
 		HAPD_CHAN_SWITCH_OTHER,
@@ -237,6 +256,12 @@ struct mxl_hostapd_data {
 #endif /* CONFIG_IEEE80211BE_MXL_MLO */
 	u8 require_csa;		/* Variable to keep track whether csa IE should be added or not */
 	u8 limit_reached;
+#ifdef CONFIG_SCS_MXL
+	struct mxl_scs_entry *scs_entry_list;
+#endif /* CONFIG_SCS_MXL */
+#ifdef CONFIG_MSCS_MXL
+	struct mxl_mscs_entry *mscs_entry_list;
+#endif /* CONFIG_MSCS_MXL */
 };
 
 int mxl_hostapd_drv_get_vendor_data(struct hostapd_data *hapd,
diff --git a/src/ap/mxl_ieee802_11.c b/src/ap/mxl_ieee802_11.c
index f67d46b4d..29d597a03 100644
--- a/src/ap/mxl_ieee802_11.c
+++ b/src/ap/mxl_ieee802_11.c
@@ -1393,6 +1393,12 @@ void mxl_assoc_req_mscs_add(struct hostapd_data *hapd, const struct ieee80211_mg
 	u8 *pos;
 	u8 len;
 
+	/* check whether station mac is in mscs_deny_mac_list */
+	if (is_mxl_qos_deny_mac(hapd, mgmt->sa, false)) {
+		wpa_printf(MSG_DEBUG, "MSCS: station mac "MACSTR "denied", MAC2STR(mgmt->sa));
+		wpa_msg(hapd->msg_ctx, MSG_INFO, "AP-STA-MSCS-DENIED "MACSTR, MAC2STR(mgmt->sa));
+		return;
+	}
 	if ((sta) && (sta->mxl_sta_info.sta_mscs_ie.mscs_descriptor) &&
 			(sta->mxl_sta_info.sta_mscs_ie.mscs_descriptor_len)&&
 			(resp == WLAN_STATUS_SUCCESS)) {
@@ -1410,4 +1416,59 @@ void mxl_assoc_req_mscs_add(struct hostapd_data *hapd, const struct ieee80211_mg
 	}
 }
 #endif /* CONFIG_MSCS_MXL */
+
+#if defined(CONFIG_SCS_MXL) || defined(CONFIG_MSCS_MXL)
+/* Add scs and mscs rules if present after sta connects 
+ * is_scs - true - add_scs_rules
+ * is_scs - false - add_mscs_rules
+ * */
+int mxl_qos_add_pending_rule(struct hostapd_data *hapd, u8 sta_addr[ETH_ALEN], bool is_scs)
+{
+	
+	int rule = 0;
+	const struct ieee80211_mgmt *mgmt = NULL;
+	struct mxl_scs_entry *scs_entry = NULL;
+	struct mxl_mscs_entry *mscs_entry = NULL;
+
+	/* Add scs rules here*/
+	if (is_scs) {
+		scs_entry = hapd->mxl_data.scs_entry_list;
+		while (scs_entry) {
+			wpa_printf(MSG_DEBUG, "STA MAC "MACSTR" || ENTRY MAC "MACSTR, MAC2STR(sta_addr), MAC2STR(scs_entry->sta_mac));
+			if (os_memcmp(scs_entry->sta_mac, sta_addr, ETH_ALEN) == 0) {
+				if (scs_entry->mgmt_bytes) {
+					mgmt =  (const struct ieee80211_mgmt *)scs_entry->mgmt_bytes;
+					mxl_handle_scs_req(hapd, mgmt, scs_entry->mgmt_len, true);
+					rule++;
+				}
+				else {
+					wpa_printf(MSG_ERROR, "SCS Rules mgmt_bytes null");
+				}
+			}
+			scs_entry = scs_entry->next;
+		}
+		wpa_printf(MSG_DEBUG, "%d MSCS Rules Added after sta connection", rule);
+	}
+	else {
+		/* Add mscs rules here */
+		mscs_entry = hapd->mxl_data.mscs_entry_list;
+		while (mscs_entry) {
+			wpa_printf(MSG_DEBUG, "STA MAC "MACSTR" || ENTRY MAC "MACSTR, MAC2STR(sta_addr), MAC2STR(mscs_entry->sta_mac));
+			if (os_memcmp(mscs_entry->sta_mac, sta_addr, ETH_ALEN) == 0) {
+				if (mscs_entry->mgmt_bytes) {
+					mgmt =  (const struct ieee80211_mgmt *)mscs_entry->mgmt_bytes;
+					mxl_handle_mscs_req(hapd, mgmt, mscs_entry->mgmt_len, true);
+					rule++;
+				}
+				else {
+					wpa_printf(MSG_ERROR, "MSCS Rules mgmt_bytes null");
+				}
+			}
+			mscs_entry = mscs_entry->next;
+		}
+		wpa_printf(MSG_DEBUG, "%d MSCS Rules Added after sta connection", rule);
+	}
+	return rule;
+}
+#endif /* CONFIG_SCS_MXL  || CONFIG_MSCS_MXL */
 #endif /* CONFIG_NATIVE_WINDOWS */
diff --git a/src/ap/mxl_ieee802_11.h b/src/ap/mxl_ieee802_11.h
index d8b4ebb30..a434018fa 100644
--- a/src/ap/mxl_ieee802_11.h
+++ b/src/ap/mxl_ieee802_11.h
@@ -74,6 +74,9 @@ u8 * mxl_hostapd_eid_mscs(struct hostapd_data *hapd, u8 *eid, size_t len, u16 st
 void mxl_assoc_req_mscs_add(struct hostapd_data *hapd, const struct ieee80211_mgmt *mgmt, struct sta_info *sta, int resp);
 void mxl_set_sta_info_mscs_descriptor(struct hostapd_data *hapd, struct sta_info *sta, const u8* elems_mscs_desc, size_t elems_mscs_len);
 #endif /* CONFIG_MSCS_MXL */
+#if defined(CONFIG_SCS_MXL) || defined(CONFIG_MSCS_MXL)
+int mxl_qos_add_pending_rule(struct hostapd_data *hapd, u8 sta_addr[ETH_ALEN], bool is_scs);
+#endif /* CONFIG_SCS_MXL  || CONFIG_MSCS_MXL */
 
 void mxl_ap_free_sta(struct hostapd_data *hapd, struct sta_info *sta);
 #endif /* MXL_IEEE802_11_H */
diff --git a/src/ap/mxl_wmm.c b/src/ap/mxl_wmm.c
index ec458cafc..dfd9c9356 100644
--- a/src/ap/mxl_wmm.c
+++ b/src/ap/mxl_wmm.c
@@ -171,6 +171,229 @@ void mxl_hostapd_eht_protected(struct hostapd_data *hapd,
 }
 #endif /* CONFIG_EPCS_MXL */
 
+#if defined(CONFIG_SCS_MXL) || defined(CONFIG_MSCS_MXL)
+/* Parser function to convert MAC hexbytes to string*/
+static void mxl_parse_mac_hex_to_str(const u8 mac[ETH_ALEN], char out[18])
+{
+	os_snprintf(out, 18, "%02x:%02x:%02x:%02x:%02x:%02x",
+			mac[0], mac[1], mac[2], mac[3], mac[4], mac[5]);
+}
+
+/* function to check qos deny mac list 
+ * is_scs - true - then scs_deny_mac_list
+ * is_scs - false - then mscs_deny_mac_list
+ * */
+int is_mxl_qos_deny_mac(struct hostapd_data *hapd, const u8 *mac, bool is_scs)
+{
+	int res = 1;
+	struct mxl_qos_deny_list *ref;
+	/* is_scs is true - scs, is_scs is false - mscs */
+	if (is_scs)
+#ifdef CONFIG_SCS_MXL
+		ref = hapd->iface->conf->mxl_conf.scs_deny;
+#endif /* CONFIG_SCS_MXL */
+	else
+#ifdef CONFIG_MSCS_MXL
+		ref = hapd->iface->conf->mxl_conf.mscs_deny;
+#endif /* CONFIG_MSCS_MXL */
+
+	while (ref) {
+		if (os_memcmp(mac, ref->mxl_qos_deny_mac, ETH_ALEN) == 0) {
+			return res;
+		}
+		ref = ref->next;
+	}
+	res = 0;
+	return res;
+}
+
+/* helper to free qos deny mac list */
+void mxl_qos_deny_mac_free(struct mxl_qos_deny_list *ref)
+{
+	struct mxl_qos_deny_list *tmp = NULL;
+	while (ref) {
+		tmp = ref->next;
+		os_free(ref);
+		ref = tmp;
+	}
+	return;
+}
+
+/*
+ * mxl_get_qos_deny_mac_list
+ * fetches  mac_address from qos list puts it into qos_buf.
+ *
+ * hapd    : reference to hapd instance
+ * qos_buf : buffer to store mac address 
+ * is_scs  : true - scs mac_list, false - mscs mac_list
+ * returns : one - success and zero - failure/empty.
+ */
+int mxl_get_qos_deny_mac_list(struct hostapd_data *hapd, char *qos_buf, bool is_scs)
+{
+	int res = 0, written = 0;
+	size_t offset = 0, len = 0;
+	char mac_str[18] = {'\0'};
+	struct mxl_qos_deny_list *ref_list = NULL;
+
+	if (is_scs) {
+#ifdef CONFIG_SCS_MXL
+		ref_list = hapd->iface->conf->mxl_conf.scs_deny;
+#endif /* CONFIG_SCS_MXL */
+		if (!ref_list)
+			return res;
+	}
+	else {
+#ifdef CONFIG_MSCS_MXL
+		ref_list = hapd->iface->conf->mxl_conf.mscs_deny;
+#endif /* CONFIG_MSCS_MXL */
+		if (!ref_list)
+			return res;
+	}
+
+	len = QOS_BUFFER_MAX;
+	while (ref_list) {
+		if ((offset + 20) < len) {
+			mxl_parse_mac_hex_to_str(ref_list->mxl_qos_deny_mac, mac_str);
+			written = snprintf(qos_buf + offset, len - offset,
+					"%s\n", mac_str);
+			if (written > 0 && offset + written < len) {
+				offset += written;
+			}
+		}
+		else {
+			wpa_printf(MSG_ERROR, "Buffer size exceeded for Qos deny mac list");
+			break;
+		}
+		ref_list = ref_list->next;
+	}
+	res = 1;
+
+	return res;
+}
+
+/*
+ * mxl_set_qos_deny_mac_list
+ * Parses mac_address sepreated spaces into a list and stores it.
+ *
+ * hapd    : reference to hapd instance
+ * mac_str : list of mac address seperated by spaces 
+ * is_scs  : true - scs mac_list, false - mscs mac_list
+ * returns : one - success and zero - failure/empty.
+ */
+int mxl_set_qos_deny_mac_list(struct hostapd_data *hapd, char *mac_str, bool is_scs)
+{
+	int res = 0;
+	u8 sta_mac[ETH_ALEN] = {'\0'};
+	char *tok = NULL;
+	char *saveptr = NULL, *mac = NULL;
+	struct mxl_qos_deny_list *ref_deny_list = NULL, *tmp_deny_list = NULL;
+
+	tok = strtok_r(mac_str, " ", &saveptr);
+	if (!tok) {
+		if (is_scs) {
+#ifdef CONFIG_SCS_MXL
+			ref_deny_list = hapd->iface->conf->mxl_conf.scs_deny;
+			hapd->iface->conf->mxl_conf.scs_deny = NULL;
+#endif /* CONFIG_SCS_MXL */
+		}
+		else {
+#ifdef CONFIG_MSCS_MXL
+			ref_deny_list = hapd->iface->conf->mxl_conf.mscs_deny;
+			hapd->iface->conf->mxl_conf.mscs_deny = NULL;
+#endif /* CONFIG_MSCS_MXL */
+		}
+		mxl_qos_deny_mac_free(ref_deny_list);
+		return res;
+	}
+
+	while (tok) {
+		mac = tok;
+		if (mac[0] != '\0') {
+			if (mxl_parse_mac_str_to_hex(mac, sta_mac) != 0) {
+				wpa_printf(MSG_ERROR, "ERR: invalid mac");
+				mxl_qos_deny_mac_free(ref_deny_list);
+				return -1;
+			}
+
+			struct mxl_qos_deny_list *qos_new_list = (struct mxl_qos_deny_list *)os_malloc(sizeof(struct mxl_qos_deny_list));
+			if (!qos_new_list) {
+				wpa_printf(MSG_ERROR, "ERR: memory allocation issue");
+				mxl_qos_deny_mac_free(ref_deny_list);
+				return -1;
+			}
+			qos_new_list->next = NULL;
+			os_memcpy(qos_new_list->mxl_qos_deny_mac, sta_mac, ETH_ALEN);
+
+			if(!tmp_deny_list)
+				tmp_deny_list = qos_new_list;
+			else {
+				qos_new_list->next = tmp_deny_list;
+				tmp_deny_list = qos_new_list;
+			}
+		}
+		tok = strtok_r(NULL, " ", &saveptr);
+	}
+
+	res = 1;
+
+	if (is_scs) {
+#ifdef CONFIG_SCS_MXL
+		ref_deny_list = hapd->iface->conf->mxl_conf.scs_deny;
+		hapd->iface->conf->mxl_conf.scs_deny = tmp_deny_list;
+#endif /* CONFIG_SCS_MXL */
+	}
+	else {
+#ifdef CONFIG_MSCS_MXL
+		ref_deny_list = hapd->iface->conf->mxl_conf.mscs_deny;
+		hapd->iface->conf->mxl_conf.mscs_deny = tmp_deny_list;
+#endif /* CONFIG_MSCS_MXL */
+	}
+
+	/* free existing entry */
+	mxl_qos_deny_mac_free(ref_deny_list);
+
+	return res;
+}
+
+/* Parser function to convert hexstream to bytes
+ * Hex -> bytes (accepts spaces/newlines/colons) 
+ * returns zero on success and non-zero on failure*/
+static int mxl_convert_hex_to_bytes(const char *hex, u8 **out, size_t *outlen)
+{
+	size_t n = 0;
+	char *clean = os_malloc(os_strlen(hex) + 1);
+	if (!clean)
+		return -1;
+
+	int clean_len = os_strlen(hex);
+	for (size_t i = 0; hex[i]; ++i) {
+		if (isxdigit((unsigned char) hex[i]) && n < clean_len) {
+			clean[n++] = hex[i];
+		}
+	}
+	clean[n] = '\0';
+
+	if (n < 2 || (n % 2) != 0) {
+		os_free(clean);
+		return -2;
+	}
+
+	*outlen = n / 2;
+	*out = os_malloc(*outlen);
+	if (!*out) {
+		os_free(clean);
+		return -3;
+	}
+
+	for (size_t i = 0; i < *outlen; ++i) {
+		char bh[3] = { clean[2*i], clean[2*i+1], 0 };
+		(*out)[i] = (u8) strtoul(bh, NULL, 16);
+	}
+	os_free(clean);
+	return 0;
+}
+#endif /* CONFIG_SCS_MXL  || CONFIG_MSCS_MXL */
+
 #ifdef CONFIG_SCS_MXL
 void mxl_send_scs_resp(struct hostapd_data *hapd,
 		const u8 *addr, struct mxl_scs_status_list *status_list, u8 scs_count, u8 dialog_token)
@@ -459,6 +682,7 @@ int mxl_scs_add_req(struct hostapd_data *hapd, const struct ieee80211_mgmt *mgmt
 			return WLAN_STATUS_UNSPECIFIED_FAILURE;
 		}
 
+		wpa_printf(MSG_INFO, "SCS: mxl_scs_add_req - eid %d", eid);
 		switch (eid) {
 		case WLAN_EID_INTRA_ACCESS_CATEGORY_PRIORITY:
 			scs_add_req.intraAccessPriority = *pos;
@@ -539,7 +763,7 @@ int mxl_scs_remove_req(struct hostapd_data *hapd,
 }
 
 void mxl_handle_scs_req(struct hostapd_data *hapd,
-			const struct ieee80211_mgmt *mgmt, size_t len)
+			const struct ieee80211_mgmt *mgmt, size_t len, bool northbound)
 {
 	int status;
 	const u8 *end;
@@ -558,6 +782,16 @@ void mxl_handle_scs_req(struct hostapd_data *hapd,
 		wpa_printf(MSG_ERROR, "SCS: mxl_scs_add_req - station not connected");
 		return;
 	}
+
+	if (!northbound) {
+		/* if request is from statio check whether sta_mac is in scs_deny_mac_list */
+		if (is_mxl_qos_deny_mac(hapd, mgmt->sa, true)) {
+			wpa_printf(MSG_DEBUG, "SCS: station mac "MACSTR "denied", MAC2STR(mgmt->sa));
+			wpa_msg(hapd->msg_ctx, MSG_INFO, "AP-STA-SCS-DENIED "MACSTR, MAC2STR(mgmt->sa));
+			return;
+		}
+	}
+
 	if (sta->eht_capab)
 		len = len - 8;
 
@@ -577,24 +811,45 @@ void mxl_handle_scs_req(struct hostapd_data *hapd,
 		switch(reqType)
 		{
 			case SCS_REQ_ADD:
+				wpa_printf(MSG_DEBUG, "SCS: mxl_handle_scs_req ADD - SCSID %u reqType %d",scsId, reqType);
 				status = mxl_scs_add_req(hapd, mgmt, pos, elen, scsId);
-				status_list[scs_count].scsId = scsId;
-				status_list[scs_count].status = status;
-				scs_count ++;
+				if (!northbound) {
+					status_list[scs_count].scsId = scsId;
+					status_list[scs_count].status = status;
+					scs_count ++;
+				}
+				else {
+					if (status == WLAN_STATUS_UNSPECIFIED_FAILURE)	
+						wpa_msg(hapd->msg_ctx, MSG_INFO, "AP-STA-SCS-ADD-PEND "MACSTR" %u", MAC2STR(mgmt->sa), scsId);
+				}
 				break;
 			case SCS_REQ_REMOVE:
+				wpa_printf(MSG_DEBUG, "SCS: mxl_handle_scs_req REMOVE - SCSID %u reqType %d",scsId, reqType);
 				status = mxl_scs_remove_req(hapd, mgmt->sa, elen, scsId);
-				status_list[scs_count].scsId = scsId;
-				status_list[scs_count].status = status;
-				scs_count ++;
+				if (!northbound) {
+					status_list[scs_count].scsId = scsId;
+					status_list[scs_count].status = status;
+					scs_count ++;
+				}
+				else {
+					if (status == WLAN_STATUS_UNSPECIFIED_FAILURE)	
+						wpa_msg(hapd->msg_ctx, MSG_INFO, "AP-STA-SCS-REM-PEND "MACSTR" %u", MAC2STR(mgmt->sa), scsId);
+				}
 				break;
 			case SCS_REQ_CHANGE:
+				wpa_printf(MSG_DEBUG, "SCS: mxl_handle_scs_req CHANGE - SCSID %u reqType %d",scsId, reqType);
 				status = mxl_scs_remove_req(hapd, mgmt->sa, elen, scsId);
 				if (status == WLAN_STATUS_TCLAS_PROCESSING_TERMINATED)
 					status = mxl_scs_add_req(hapd, mgmt, pos, elen, scsId);
-				status_list[scs_count].scsId = scsId;
-				status_list[scs_count].status = status;
-				scs_count ++;
+				if (!northbound) {
+					status_list[scs_count].scsId = scsId;
+					status_list[scs_count].status = status;
+					scs_count ++;
+				}
+				else {
+					if (status == WLAN_STATUS_UNSPECIFIED_FAILURE)	
+						wpa_msg(hapd->msg_ctx, MSG_INFO, "AP-STA-SCS-UP-PEND "MACSTR" %u", MAC2STR(mgmt->sa), scsId);
+				}
 				break;
 			default:
 				hostapd_logger(hapd, mgmt->sa, HOSTAPD_MODULE_IEEE80211,
@@ -606,7 +861,198 @@ void mxl_handle_scs_req(struct hostapd_data *hapd,
 
 		pos += elen;
 	}
-	mxl_send_scs_resp(hapd, mgmt->sa, status_list, scs_count, dialog_token);
+	if (!northbound)
+		mxl_send_scs_resp(hapd, mgmt->sa, status_list, scs_count, dialog_token);
+}
+
+/*
+ * mxl_northbound_scs_list_req
+ * get SCS rule from scs_data_entry list and store to buf
+ *
+ * hapd         : ref to hapd instance
+ * scs_list_buf : store list entries
+ * returns      : one - success and zero - failure or empty.
+ */
+int mxl_northbound_scs_list_rule(struct hostapd_data *hapd, char *scs_list_buf)
+{
+	int res = 0;
+	size_t len = 0;
+	int offset = 0, written = 0;
+	char mac_str[18] = {'\0'};
+	struct mxl_scs_entry *ref_entry = NULL;
+
+	ref_entry = hapd->mxl_data.scs_entry_list;
+	if (!ref_entry)
+		return res;
+	else
+		res = 1;
+
+	/* logic to loop through each entry and store data in buf */
+	len = QOS_BUFFER_MAX;
+	while (ref_entry) {
+		if ((offset + 21) < len) {
+			mxl_parse_mac_hex_to_str(ref_entry->sta_mac, mac_str);
+			written = snprintf(scs_list_buf + offset, len - offset,
+					"%u %s\n",ref_entry->scsId, mac_str);
+
+			if (written > 0 && offset + written < len) {
+				offset += written;
+			}
+		}
+		else {
+			wpa_printf(MSG_ERROR, "Buffer size exceeds for SCS rule listing");
+			break;
+		}
+		ref_entry = ref_entry->next;
+	}
+
+	return res;
+}
+
+/*
+ * mxl_scs_remove_entry
+ * Remove SCS rule from northbound API in scs_data_entry list
+ *
+ * hapd     : ref to hapd instance
+ * sta_mac  : mac address of station
+ * scsId    : scsId of the rule to be added
+ * returns  : one - success and zero - failure.
+ */
+int mxl_scs_remove_entry(struct hostapd_data *hapd, const u8 sta_mac[ETH_ALEN], unsigned int scsId)
+{
+	int res = 0;
+	struct mxl_scs_entry *ref_entry = hapd->mxl_data.scs_entry_list; 
+	struct mxl_scs_entry *prev_entry = NULL;
+
+	/* logic to remove entry on matching scs and sta mac */
+	while (ref_entry) {
+		if (os_memcmp(ref_entry->sta_mac, sta_mac, ETH_ALEN) == 0 && 
+			ref_entry->scsId == scsId) {
+
+			if (prev_entry) {
+				prev_entry->next = ref_entry->next;
+			}
+			else {
+				hapd->mxl_data.scs_entry_list = ref_entry->next;
+			}
+
+			os_free(ref_entry->mgmt_bytes);
+			os_free(ref_entry);
+			res = 1;
+			break;
+		}
+		prev_entry = ref_entry;
+		ref_entry = ref_entry->next;
+	}
+
+	return res;
+}
+
+/*
+ * mxl_northbound_scs_remove_req
+ * SCS remove request from northbound API is handled where copy in scs_data_entry
+ * is removed and  handle_scs_req to remove rule from driver.
+ *
+ * mac_str : mac address of station in string format 
+ * scsId   : scsId of the rule to be removed
+ * returns : one - success and zero - failure.
+ */
+int mxl_northbound_scs_remove_req(struct hostapd_data *hapd, char *mac_str, unsigned int scsId)
+{
+	int res = 0;
+	u8 mac[ETH_ALEN] = {'\0'};
+
+	if (mxl_parse_mac_str_to_hex(mac_str, mac) != 0) {
+		wpa_printf(MSG_ERROR, "ERR: invalid STA_MAC\n");
+		return res;
+	}
+
+	/* remove scs entry from mxl_scs_entry list */
+	res = mxl_scs_remove_entry(hapd, mac, scsId);
+	if (!res) {
+		wpa_printf(MSG_ERROR, "ERR: failed to remove scs entry\n");
+		return res;	
+	}
+
+	/* scs remove request hook to driver */
+	mxl_scs_remove_req(hapd, mac, 0, scsId);
+
+	return res;
+}
+
+/*
+ * mxl_northbound_scs_add_req
+ * Parses SCS request from northbound API and stores a copy in scs_data_entry
+ * as well as calls handle_scs_req to add to driver.
+ *
+ * mac_str : mac address of station in string format 
+ * scsId   : scsId of the rule to be added
+ * hex_str : scs req in hex format
+ * returns : one - success and zero - failure.
+ */
+int mxl_northbound_scs_add_req(struct hostapd_data *hapd, char *mac_str, unsigned int scsId, char *hex_str)
+{
+	int res = 0;
+	u8 mac[ETH_ALEN] = {'\0'};
+	u8 *mgmt_buf = NULL;
+	size_t mgmt_len = 0;
+	const struct ieee80211_mgmt *mgmt = NULL;
+	struct mxl_scs_entry *ref_entry = hapd->mxl_data.scs_entry_list; 
+	struct mxl_scs_entry *new_entry = NULL;
+
+	if (mxl_parse_mac_str_to_hex(mac_str, mac) != 0) {
+		wpa_printf(MSG_ERROR, "ERR: invalid STA_MAC\n");
+		return res;
+	}
+
+	if (mxl_convert_hex_to_bytes(hex_str, &mgmt_buf, &mgmt_len)) {
+		wpa_printf(MSG_ERROR, "ERR: invalid hexstream\n");
+		return res;
+	}
+
+	/* typecast to mgmt structure */
+	mgmt = (const struct ieee80211_mgmt *) mgmt_buf;
+
+	new_entry = (struct mxl_scs_entry *)os_malloc(sizeof(struct mxl_scs_entry));
+	if (!new_entry) {
+		wpa_printf(MSG_ERROR, "Memory allocation issue in SCS store");
+		os_free(mgmt_buf);
+		return res;
+	}
+
+	/* Avoid redundancy in SCS req */
+	if (!mxl_scs_remove_entry(hapd, mac, scsId)) {
+		wpa_printf(MSG_DEBUG, "No existing SCS entry before updating");
+	}
+
+	/* Copy params to scs list */
+	new_entry->scsId = scsId;
+	os_memcpy(new_entry->sta_mac, mac, ETH_ALEN);
+	new_entry->mgmt_bytes = os_malloc(mgmt_len);
+	if (!new_entry->mgmt_bytes) { 
+		wpa_printf(MSG_ERROR, "Memory allocation issue in SCS store");
+		os_free(new_entry);
+		os_free(mgmt_buf);
+		return res;
+	}
+	os_memcpy(new_entry->mgmt_bytes, mgmt, mgmt_len);
+	new_entry->mgmt_len = mgmt_len;
+	new_entry->next = NULL;
+	res = 1;
+
+	/* preserve scs list nodes */
+	if(!ref_entry)
+		ref_entry = new_entry;
+	else {
+		new_entry->next = ref_entry;
+		ref_entry = new_entry;
+	}
+
+	hapd->mxl_data.scs_entry_list = ref_entry;
+	/* scs add request hook to driver */
+	mxl_handle_scs_req(hapd, mgmt, mgmt_len, true);
+
+	return res;
 }
 #endif /* CONFIG_SCS_MXL */
 
@@ -757,10 +1203,11 @@ int mxl_mscs_remove_req(struct hostapd_data *hapd,
 }
 
 void mxl_handle_mscs_req(struct hostapd_data *hapd,
-			const struct ieee80211_mgmt *mgmt, size_t len)
+			const struct ieee80211_mgmt *mgmt, size_t len, bool northbound)
 {
 	int status;
 	u8 dialog_token;
+	struct sta_info *sta;
 	const u8 *pos = mgmt->u.action.u.mscs_action_req.variable;
 	u8 req_type, ext_eid, eid, elen;
 
@@ -769,6 +1216,21 @@ void mxl_handle_mscs_req(struct hostapd_data *hapd,
 		return;
 	}
 
+	sta = ap_get_sta(hapd, mgmt->sa);
+	if (sta == NULL) {
+		wpa_printf(MSG_ERROR, "MSCS: mxl_mscs_add_req - station not connected");
+		return;
+	}
+
+	if (!northbound) {
+		/* if request is from station check whether sta_mac is in mscs_deny_mac_list */
+		if (is_mxl_qos_deny_mac(hapd, mgmt->sa, false)) {
+			wpa_printf(MSG_DEBUG, "MSCS: station mac "MACSTR "denied", MAC2STR(mgmt->sa));
+			wpa_msg(hapd->msg_ctx, MSG_INFO, "AP-STA-MSCS-DENIED "MACSTR, MAC2STR(mgmt->sa));
+			return;
+		}
+	}
+
 	eid = *pos++;
 	if(eid != WLAN_EID_EXTENSION) {
 	  wpa_printf(MSG_ERROR, "MSCS: mxl_handle_mscs_req - MSCS Descriptor element is not present");
@@ -792,17 +1254,32 @@ void mxl_handle_mscs_req(struct hostapd_data *hapd,
 	{
 		case MSCS_REQ_ADD:
 			status = mxl_mscs_add_req(hapd, mgmt, pos, elen);
-			mxl_send_mscs_resp(hapd, mgmt->sa, status, dialog_token);
+			if (!northbound)
+				mxl_send_mscs_resp(hapd, mgmt->sa, status, dialog_token);
+			else {
+				if (status == WLAN_STATUS_UNSPECIFIED_FAILURE)
+					wpa_msg(hapd->msg_ctx, MSG_INFO, "AP-STA-MSCS-ADD-PEND "MACSTR, MAC2STR(mgmt->sa));
+			}
 			break;
 		case MSCS_REQ_REMOVE:
 			status = mxl_mscs_remove_req(hapd, mgmt->sa, elen);
-			mxl_send_mscs_resp(hapd, mgmt->sa, status, dialog_token);
+			if (!northbound)
+				mxl_send_mscs_resp(hapd, mgmt->sa, status, dialog_token);
+			else {
+				if (status == WLAN_STATUS_UNSPECIFIED_FAILURE)
+					wpa_msg(hapd->msg_ctx, MSG_INFO, "AP-STA-MSCS-REM-PEND "MACSTR, MAC2STR(mgmt->sa));
+			}
 			break;
 		case MSCS_REQ_CHANGE:
 			status = mxl_mscs_remove_req(hapd, mgmt->sa, elen);
 			if (status == WLAN_STATUS_TCLAS_PROCESSING_TERMINATED)
 				status = mxl_mscs_add_req(hapd, mgmt, pos, elen);
-			mxl_send_mscs_resp(hapd, mgmt->sa, status, dialog_token);
+			if (!northbound)
+				mxl_send_mscs_resp(hapd, mgmt->sa, status, dialog_token);
+			else {
+				if (status == WLAN_STATUS_UNSPECIFIED_FAILURE)
+					wpa_msg(hapd->msg_ctx, MSG_INFO, "AP-STA-MSCS-UP-PEND "MACSTR, MAC2STR(mgmt->sa));
+			}
 			break;
 		default:
 			hostapd_logger(hapd, mgmt->sa, HOSTAPD_MODULE_IEEE80211,
@@ -812,6 +1289,175 @@ void mxl_handle_mscs_req(struct hostapd_data *hapd,
 			return;
 	}
 }
+
+/*
+ * mxl_northbound_mscs_list_req
+ * get MSCS rule from mscs_data_entry list and store to buf
+ *
+ * hapd          : ref to hapd instance
+ * mscs_list_buf : store list entries
+ * returns       : one - success and zero - failure or empty.
+ */
+int mxl_northbound_mscs_list_rule(struct hostapd_data *hapd, char *mscs_list_buf)
+{
+	int res = 0;
+	size_t len = 0;
+	int offset = 0, written = 0;
+	char mac_str[18] = {'\0'};
+	struct mxl_mscs_entry *ref_entry = NULL;
+
+	ref_entry = hapd->mxl_data.mscs_entry_list;
+	if (!ref_entry)
+		return res;
+	else
+		res = 1;
+
+	/* logic to loop through each entry and store data in buf */
+	len = QOS_BUFFER_MAX;
+	while (ref_entry) {
+		if ((offset + 21) < len) {
+			mxl_parse_mac_hex_to_str(ref_entry->sta_mac, mac_str);
+			written = snprintf(mscs_list_buf + offset, len - offset,
+                  "%s\n", mac_str);
+
+			if (written > 0 && offset + written < len) {
+				offset += written;
+			}
+		}
+		else {
+			wpa_printf(MSG_ERROR, "Buffer size exceeds for MSCS rule listing");
+			break;
+		}
+		ref_entry = ref_entry->next;
+	}
+
+	return res;
+}
+
+/*
+ * mxl_northbound_mscs_remove_req
+ * MSCS remove request from northbound API is handled where copy in mscs_data_entry
+ * is removed and  handle_mscs_req to remove rule from driver.
+ *
+ * mac_str : mac address of station in string format 
+ * returns : one - success and zero - failure.
+ */
+int mxl_northbound_mscs_remove_req(struct hostapd_data *hapd, char *mac_str)
+{
+	int res = 0, match = 0;
+	u8 mac[ETH_ALEN] = {'\0'};
+	struct mxl_mscs_entry *ref_entry = hapd->mxl_data.mscs_entry_list; 
+	struct mxl_mscs_entry *prev_entry = NULL;
+
+	if (mxl_parse_mac_str_to_hex(mac_str, mac) != 0) {
+		wpa_printf(MSG_ERROR, "ERR: invalid STA_MAC\n");
+		return res;
+	}
+
+	/* logic to remove entry on matching mscs and sta mac */
+	while (ref_entry) {
+		match = 0;
+		/* if match found free that entry*/
+		if (os_memcmp(ref_entry->sta_mac, mac, ETH_ALEN) == 0) {
+			if (prev_entry) {
+				prev_entry->next = ref_entry->next;
+			}
+			else {
+				hapd->mxl_data.mscs_entry_list = ref_entry->next;
+			}
+
+			os_free(ref_entry->mgmt_bytes);
+			os_free(ref_entry);
+			/* reset the node to continue through the list */
+			if (prev_entry)
+				ref_entry = prev_entry->next;
+			else {
+				ref_entry = hapd->mxl_data.mscs_entry_list;
+			}
+			match = 1;
+			res++;
+		}
+		if (!match) {
+			prev_entry = ref_entry;
+			ref_entry = ref_entry->next;
+		}
+	}
+
+	/* mscs remove request hook to driver */
+	mxl_mscs_remove_req(hapd, mac, 0);
+
+	return res;
+}
+
+/*
+ * mxl_northbound_mscs_add_req
+ * Parses MSCS request from northbound API and stores a copy in mscs_data_entry
+ * as well as calls handle_mscs_req to add to driver.
+ *
+ * mac_str : mac address of station in string format 
+ * hex_str : mscs req in hex format
+ * returns : one - success and zero - failure.
+ */
+int mxl_northbound_mscs_add_req(struct hostapd_data *hapd, char *mac_str, char *hex_str)
+{
+	int res = 0;
+	u8 mac[ETH_ALEN] = {'\0'};
+	u8 *mgmt_buf = NULL;
+	size_t mgmt_len = 0;
+	const struct ieee80211_mgmt *mgmt = NULL;
+	struct mxl_mscs_entry *ref_entry = hapd->mxl_data.mscs_entry_list; 
+	struct mxl_mscs_entry *new_entry = NULL;
+
+	if (mxl_parse_mac_str_to_hex(mac_str, mac) != 0) {
+		wpa_printf(MSG_ERROR, "ERR: invalid STA_MAC\n");
+		return res;
+	}
+
+	if (mxl_convert_hex_to_bytes(hex_str, &mgmt_buf, &mgmt_len)) {
+		wpa_printf(MSG_ERROR, "ERR: invalid hexstream\n");
+		return res;
+	}
+
+	/* typecast to mgmt structure */
+	mgmt = (const struct ieee80211_mgmt *) mgmt_buf;
+
+	/* store mgmt in mscs_entry_list */
+	new_entry = (struct mxl_mscs_entry *)os_malloc(sizeof(struct mxl_mscs_entry));
+	if (!new_entry) {
+		wpa_printf(MSG_ERROR, "Memory allocation issue in MSCS store");
+		os_free(mgmt_buf);
+		return res;
+	}
+
+	/*TODO : Decide whether MSCS req has to checked fore redundnacy */
+	/* Copy params to mscs list */
+	os_memcpy(new_entry->sta_mac, mac, ETH_ALEN);
+	new_entry->mgmt_bytes = os_malloc(mgmt_len);
+	if (!new_entry->mgmt_bytes) { 
+		wpa_printf(MSG_ERROR, "Memory allocation issue in MSCS store");
+		os_free(mgmt_buf);
+		os_free(new_entry);
+		return res;
+	}
+	os_memcpy(new_entry->mgmt_bytes, mgmt, mgmt_len);
+	new_entry->mgmt_len = mgmt_len;
+	new_entry->next = NULL;
+	res = 1;
+
+	/* preserve mscs list nodes */
+	if(!ref_entry)
+		ref_entry = new_entry;
+	else {
+		new_entry->next = ref_entry;
+		ref_entry = new_entry;
+	}
+
+	hapd->mxl_data.mscs_entry_list = ref_entry;
+	/* mscs add request hook to driver */
+	mxl_handle_mscs_req(hapd, mgmt, mgmt_len, true);
+
+	return res;
+}
 #endif /* CONFIG_MSCS_MXL */
 
 #if defined(CONFIG_SCS_MXL) || defined(CONFIG_MSCS_MXL)
@@ -824,13 +1470,13 @@ void mxl_hostapd_handle_robust_av_stream(struct hostapd_data *hapd,
 	case ROBUST_AV_SCS_REQ:
 #ifdef CONFIG_SCS_MXL
 		wpa_printf(MSG_DEBUG, "mxl_hostapd_scs_action: ROBUST_AV_SCS_REQ");
-		mxl_handle_scs_req(hapd, mgmt, len);
+		mxl_handle_scs_req(hapd, mgmt, len, false);
 #endif /* CONFIG_SCS_MXL */
 		break;
 	case ROBUST_AV_MSCS_REQ:
 #ifdef CONFIG_MSCS_MXL
 		wpa_printf(MSG_DEBUG, "mxl_hostapd_mscs_action: ROBUST_AV_MSCS_REQ");
-		mxl_handle_mscs_req(hapd, mgmt, len);
+		mxl_handle_mscs_req(hapd, mgmt, len, false);
 #endif /* CONFIG_MSCS_MXL */
 		break;
 default:
diff --git a/src/ap/mxl_wmm.h b/src/ap/mxl_wmm.h
index a1d3bf9ea..d90898b84 100644
--- a/src/ap/mxl_wmm.h
+++ b/src/ap/mxl_wmm.h
@@ -25,24 +25,34 @@ void mxl_hostapd_eht_protected(struct hostapd_data *hapd,
 #endif /* CONFIG_EPCS_MXL */
 
 #if defined(CONFIG_SCS_MXL) || defined(CONFIG_MSCS_MXL)
+#define QOS_BUFFER_MAX 2048
 void mxl_hostapd_handle_robust_av_stream(struct hostapd_data *hapd, const struct ieee80211_mgmt *mgmt, size_t len);
+int is_mxl_qos_deny_mac(struct hostapd_data *hapd, const u8 *mac, bool is_scs);
+int mxl_get_qos_deny_mac_list(struct hostapd_data *hapd, char *qos_buf, bool is_scs);
+int mxl_set_qos_deny_mac_list(struct hostapd_data *hapd, char *mac_str, bool is_scs);
 #endif /* CONFIG_SCS_MXL || CONFIG_MSCS_MXL */
 
 #ifdef CONFIG_SCS_MXL
-void mxl_handle_scs_req(struct hostapd_data *hapd, const struct ieee80211_mgmt *mgmt, size_t len);
+void mxl_handle_scs_req(struct hostapd_data *hapd, const struct ieee80211_mgmt *mgmt, size_t len, bool northbound);
 int mxl_scs_add_req(struct hostapd_data *hapd, const struct ieee80211_mgmt *mgmt, const u8 *pos, size_t len, u8 scsId);
 int mxl_scs_remove_req(struct hostapd_data *hapd, const u8 *addr, size_t len, u8 scsId);
 int mxl_scs_parse_qos_characteristics_elems(const u8 *pos, struct mxl_scs_add_req *scs_add_req);
 int mxl_scs_parse_tclas_elems(const u8 *pos, struct mxl_scs_add_req *scs_add_req, size_t len, u8 *elems_num);
 void mxl_send_scs_resp(struct hostapd_data *hapd, const u8 *addr, struct mxl_scs_status_list *status_list, u8 scs_count, u8 dialog_token);
+int mxl_northbound_scs_list_rule(struct hostapd_data *hapd, char *scs_list_buf);
+int mxl_northbound_scs_remove_req(struct hostapd_data *hapd, char *mac_str, unsigned int scsId);
+int mxl_northbound_scs_add_req(struct hostapd_data *hapd, char *mac_str, unsigned int scsId, char *hex_str);
 #endif /* CONFIG_SCS_MXL */
 
 #ifdef CONFIG_MSCS_MXL
-void mxl_handle_mscs_req(struct hostapd_data *hapd, const struct ieee80211_mgmt *mgmt, size_t len);
+void mxl_handle_mscs_req(struct hostapd_data *hapd, const struct ieee80211_mgmt *mgmt, size_t len, bool northbound);
 int mxl_mscs_add_req(struct hostapd_data *hapd, const struct ieee80211_mgmt *mgmt, const u8 *pos, size_t len);
 int mxl_mscs_remove_req(struct hostapd_data *hapd, const u8 *addr, size_t len);
 int mxl_mscs_parse_tclas_mask_elems(const u8 *pos, struct mxl_mscs_add_req *mscs_add_req);
 void mxl_send_mscs_resp(struct hostapd_data *hapd, const u8 *addr, int status, u8 dialog_token);
+int mxl_northbound_mscs_list_rule(struct hostapd_data *hapd, char *mscs_list_buf);
+int mxl_northbound_mscs_remove_req(struct hostapd_data *hapd, char *mac_str);
+int mxl_northbound_mscs_add_req(struct hostapd_data *hapd, char *mac_str, char *hex_str);
 #endif /* CONFIG_MSCS_MXL */
 #endif /* MXL_WME_H */
 
diff --git a/src/ap/sta_info.c b/src/ap/sta_info.c
index c25abd3d1..570bb7be8 100644
--- a/src/ap/sta_info.c
+++ b/src/ap/sta_info.c
@@ -2589,6 +2589,19 @@ void ap_sta_set_authorized_event(struct hostapd_data *hapd,
 		wpa_msg(hapd->msg_ctx, MSG_INFO, AP_STA_CONNECTED "%s%s%s%s%s%s%s%s%s",
 			buf, ip_addr, signal_strength, rates, capabilities, assoc_req, sta_extra_cap, keyid_buf, dpp_pkhash_buf);
 
+#ifdef CONFIG_SCS_MXL
+		/* Add northbound scs_rules upon sta connection */
+		if (mxl_qos_add_pending_rule(hapd, sta->addr, true)) {
+			wpa_printf(MSG_DEBUG, "SCS Pending Rules Added for sta " MACSTR, MAC2STR(sta->addr));
+		}
+#endif /* CONFIG_SCS_MXL */
+#ifdef CONFIG_MSCS_MXL
+		/* Add northbound mscs_rules upon sta connection */
+		if (mxl_qos_add_pending_rule(hapd, sta->addr, false)) {
+			wpa_printf(MSG_DEBUG, "MSCS Pending Rules Added for sta " MACSTR, MAC2STR(sta->addr));
+		}
+#endif /* CONFIG_MSCS_MXL */
+
 		if (hapd->msg_ctx_parent &&
 		    hapd->msg_ctx_parent != hapd->msg_ctx)
 			wpa_msg_no_global(hapd->msg_ctx_parent, MSG_INFO,
@@ -2598,7 +2611,7 @@ void ap_sta_set_authorized_event(struct hostapd_data *hapd,
 		os_free(rates);
 		os_free(capabilities);
 		os_free(assoc_req);
-#else
+#else /* CONFIG_VENDOR_MXL */
 		wpa_msg(hapd->msg_ctx, MSG_INFO, AP_STA_CONNECTED "%s%s%s%s",
 			buf, ip_addr, keyid_buf, dpp_pkhash_buf);
 
-- 
2.43.0

